### PR TITLE
groovy2kotlin pact-jvm-model: MatchingRulesImpl

### PR DESCRIPTION
**Pull request by tmszdmsk**
Related to #439.

Since there is less Groovy than Scala code that we want to replace with Kotlin i picked `MatchingRulesImpl.groovy` as a good candidate for rewrite to Kotlin.

Notes:
* I had to change two tests that depended on specific format used by Groovy `.toString()` methods to serialize object, is anyone depending on this specific groovish format?
* I see a lot of things that are wrong with this class but this PR is about "blindly" rewriting to kotlin
* `PactSerialiserSpec` had to be changed because somehow equality comparison changed and `TypeMatcher` from Kotlin didn't equal `new TypeMatcher()` from Groovy (which is privileged to call private constructors ;)  

Disclaimer:
* I am not a Kotlin master so probably not all things are idiomatic ;)

What do you think?

### CODE REVIEW

File: pact-jvm-consumer-groovy/src/test/groovy/au/com/dius/pact/consumer/groovy/PactResultSpec.groovy [hunk 1]
```diff
@@ -134,7 +134,7 @@ class PactResultSpec extends Specification {
           '\tpath: /path\n' +
           '\tquery: [:]\n' +
           '\theaders: [Content-Type:application/json]\n' +
-          '\tmatchers: MatchingRules(rules=[body:Category(name=body, matchingRules={}), path:Category(name=path, matchingRules={})])\n' +
+          '\tmatchers: MatchingRules(rules={body=Category(name=body, matchingRules={}), path=Category(name=path, matchingRules={})})\n' +
```
Comment (tmszdmsk) [inline pact-jvm-consumer-groovy/src/test/groovy/au/com/dius/pact/consumer/groovy/PactResultSpec.groovy]: `.toString` formatting is different in kotlin.

File: pact-jvm-model/src/test/groovy/au/com/dius/pact/model/PactSerialiserSpec.groovy [hunk 2]
```diff
@@ -39,12 +39,12 @@ class PactSerialiserSpec extends Specification {
     provider = new Provider('test_provider')
     consumer = new Consumer('test_consumer')
     def requestMatchers = new MatchingRulesImpl()
-    requestMatchers.addCategory('body').addRule('$.test', new TypeMatcher())
+    requestMatchers.addCategory('body').addRule('$.test', TypeMatcher.INSTANCE)
```
Comment (tmszdmsk) [inline pact-jvm-model/src/test/groovy/au/com/dius/pact/model/PactSerialiserSpec.groovy]: calling private constructor of TypeMatcher was breaking equality test. Strange thing is that it worked before ;)

File: pact-jvm-consumer-groovy/src/test/groovy/au/com/dius/pact/consumer/groovy/PactResultSpec.groovy (hunk 1 above)
Comment (uglyog) [reply pact-jvm-consumer-groovy/src/test/groovy/au/com/dius/pact/consumer/groovy/PactResultSpec.groovy]: This does raise the issue of the output changing, which is mostly used in logging. Hopefully nobody depends on the format of the matchers which are output to the log files.

File: pact-jvm-model/src/test/groovy/au/com/dius/pact/model/PactSerialiserSpec.groovy (hunk 2 above)
Comment (uglyog) [reply pact-jvm-model/src/test/groovy/au/com/dius/pact/model/PactSerialiserSpec.groovy]: oh, yeah, that is wrong. TypeMatcher is a singleton so the constructor should have never been called, but you can call private things in groovy.

**Comment by uglyog** [comment]
All good, thank you. 

There are some improvements that can be made now that it is Kotlin source (like using a data class, which provides the copy, equals and hashcode methods by default), and cleanup some of the nested if statements. I'll do that as soon as I get some spare time.

**Pull request merged**